react/features/video-layout/subscriber.js

```diff
@@ -4,21 +4,69 @@ import {
     VIDEO_QUALITY_LEVELS,
     setMaxReceiverVideoQuality
 } from '../base/conference';
+import {
+    getPinnedParticipant,
+    pinParticipant
+} from '../base/participants';
 import { StateListenerRegistry } from '../base/redux';
 import { selectParticipant } from '../large-video';
 import { shouldDisplayTileView } from './functions';
 
+declare var interfaceConfig: Object;
+
 /**
  * StateListenerRegistry provides a reliable way of detecting changes to
  * preferred layout state and dispatching additional actions.
  */
 StateListenerRegistry.register(
     /* selector */ state => shouldDisplayTileView(state),
-    /* listener */ (displayTileView, { dispatch }) => {
-        dispatch(selectParticipant());
+    /* listener */ (displayTileView, store) => {
+        store.dispatch(selectParticipant());
 
         if (!displayTileView) {
-            dispatch(setMaxReceiverVideoQuality(VIDEO_QUALITY_LEVELS.HIGH));
+            store.dispatch(
+                setMaxReceiverVideoQuality(VIDEO_QUALITY_LEVELS.HIGH));
+
+            _updateAutoPinnedParticipant(store);
         }
     }
 );
+
+/**
+ * For auto-pin mode, listen for changes to the last known screen share
+ * participant and automatically pin that participant.
+ */
+StateListenerRegistry.register(
```

Comment:
I'm wondering is there any problem to handle this in the middleware? What are the benefits to handle it in a state listener versus the middleware?

Reply:
I did it for convenience while figuring out how to detect screenshare changes. I could use subscriber. You want me to convert? I'd likely listen for changes to the track state and iterate over each track to find out what changed, and then iterate over the known screenshares to see if there are any removals.

Reply:
@virtuacoplenny but isn't it easier to do it with the state listener ? I'd prefer a state listener here

Reply:
Oh no, don't revert it! I'm just trying to figure out when we are using the state listeners :)

Reply:
I switched over to a state listeners. Also used an array over a set to store screen share IDs just because Hristo had pointed it out; I think it's a toss-up and the perf difference, if any, is probably nothing, so I want to say it comes down to style for now as the extra code for finding isn't a big deal and its usage is limited.